src/Form/FormGroupContext.jsx

```diff
@@ -1,61 +1,82 @@
-      contextValue['aria-describedby'],
-      props['aria-describedby'],
-    ),
-    onFocus: callAllHandlers(contextValue.onFocus, props.onFocus),
-    onBlur: callAllHandlers(contextValue.onBlur, props.onBlur),
+const useStateEffect = (initialState) => {
+  const [state, setState] = useState(initialState);
+  const useSetStateEffect = (newState) => {
+    useEffect(() => setState(newState), [newState]);
   };
-  return mergedValues;
+  return [state, useSetStateEffect];
 };
 
 const FormGroupContextProvider = ({
   children,
   controlId,
   isInvalid,
   isValid,
-  onFocus,
-  onBlur,
   size,
 }) => {
   const resolvedId = React.useMemo(() => controlId || newId('form-field'), [controlId]);
-  const [hasFocus, setHasFocusTrue, setHasFocusFalse] = useToggle(false);
-  const [describedByIds, getNewDescriptorId, removeDescriptorId] = useIdList(resolvedId);
+  const [describedByIds, useRegisteredDescriptorId] = useIdList(resolvedId);
+  const [labelledByIds, useRegisteredLabellerId] = useIdList(resolvedId);
+  const [controlIsGroup, useSetControlIsGroupEffect] = useStateEffect(false);
+
+  const getControlProps = (controlProps) => {
+    // labelledByIds from the list above should only be added to a control
+    // if it the control is a group. We prefer adding a condition here because:
+    //    - Hooks cannot be called inside conditionals
+    //    - The getLabelProps function below is forced to generate an id
+    //      whether it is needed or not.
+    //    - This is what allows consumers of Paragon to use <Form.Label>
+    //      interchangeably between ControlGroup type controls and regular Controls
+    const labelledByIdsForControl = controlIsGroup ? labelledByIds : undefined;
+    return omitUndefinedProperties({
+      ...controlProps,
+      'aria-describedby': classNames(controlProps['aria-describedby'], describedByIds) || undefined,
+      'aria-labelledby': classNames(controlProps['aria-labelledby'], labelledByIdsForControl) || undefined,
+      id: resolvedId,
+    });
+  };
+
+  const getLabelProps = (labelProps) => {
+    const id = useRegisteredLabellerId(labelProps?.id);
+    if (controlIsGroup) {
+      return { ...labelProps, id };
+    }
+    return { ...labelProps, htmlFor: resolvedId };
+  };
+
+  const getDescriptorProps = (descriptorProps) => {
+    const id = useRegisteredDescriptorId(descriptorProps?.id);
+    return { ...descriptorProps, id };
+  };
+
   const contextValue = {
+    getControlProps,
+    getLabelProps,
+    getDescriptorProps,
+    useSetControlIsGroupEffect,
+    controlIsGroup,
     controlId: resolvedId,
     isInvalid,
     isValid,
     size,
-    hasFocus,
-    getNewDescriptorId,
-    removeDescriptorId,
-    'aria-describedby': describedByIds.length ? describedByIds.join(' ') : undefined,
-    onFocus: callAllHandlers(setHasFocusTrue, onFocus),
-    onBlur: callAllHandlers(setHasFocusFalse, onBlur),
   };
   return (
     <FormGroupContext.Provider value={contextValue}>
```

Comment:
Should `contextValue` be memoized with `useMemo` to avoid any unnecessary rendering of children components nested within this context provider? That way, the nested components will only get re-rendered when the context values actually change vs. any time the parent `FormGroupContextProvider` is re-rendered?

Reply:
I can add in a follow up PR if this is needed. I'll bug you about it in person